### PR TITLE
Add promise completion report and improve promise reporting verbiage

### PR DESCRIPTION
**Pull request by erikerikson**
## What did you implement

The language used to report promise status is challenging to understand and doesn't state when completion has occurred.  This PR attempts to improve this.  For reference, see [comments on #4984](https://github.com/serverless/serverless/issues/4984#issuecomment-545388840).

Improves on #4984

Progress:

```
############################################################################################
    # 2: 3 of 5 promises have settled
    # 2: 2 are taking longer than expected:
    # 2:   foo waited on by: ${foo:}
    # 2:   bar waited on by: ${bar:}
    # This can result from latent connections but may represent a cyclic variable dependency
    ##########################################################################################
```

Completion:

```
############################################################################################
    # Completed after 6ms
    # 5 promises are in the following states:
    #   resolved: 3
    #   rejected: 1
    #   pending:  1
    ##########################################################################################
```

## How can we verify it

New tests added.

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by medikoo** [comment]
Sorry I needed to revert it, explained it in revert PR: https://github.com/serverless/serverless/pull/6963

**Comment by erikerikson** [comment]
TL;DR: @medikoo thanks for catching this oversight.  The use case is silently hung variable resolution.  Scoping the final report should eliminate the noise and I will submit a PR for that.  Two questions at end.

Details:

No need to apologize, in  hindsight it was the correct decision to revert.  Sorry for my oversight.  These changes could be improved to only trigger the final report if one or more status reports were shown.  At this point I think this improvement should be made and the PR reopened.  I am happy to close it if this conversation suggests a different course of action.

To be clear, #6960 is only a messaging change and did not introduce the promise tracker.  Instead it was intended to cause its messages to be more user friendly.  That written, I think it's great to ask whether that is a component which has outlived its usefulness.  I will provide an accounting of the rationale under which it was written.

The promise tracker was introduced with a refactoring and revision of the variable resolution code to resolve a usability concern.  It is currently exclusively used there.  There were deadlock and other challenges resulting from the prior variable resolution strategy which was made asynchronous but could cause unnecessary errors and indefinite hanging.  Causes include slow remote services and improper circular references.

\<aside>
The resolution was roughly completing the resolution of variables before resolving the content they resolve to.  Consider a file that references itself and is loaded using the file variable type.  The file loading is asynchronous and results in an unsettled (not resolved or rejected) promise.  If when you are resolving the promise (because the file contents are loaded) you immediately process and resolve the contents then the self-reference will attempt to access the result of the promise associated with the loading of the file which is now in turn awaiting the resolution of the self-reference: thus a soft deadlock has occurred.  By completely resolving the file prior to resolving its contents I had to introduce a variable resolution "pause" mechanism (see the "deep" variable type) to allow that content to be revisited after the completion.  The [maintainer documentation](https://github.com/serverless/serverless/blob/master/lib/classes/Variables.js#L24) speaks about "generations" of variable resolution which means variables in the serverless config are resolved over and over [until there are no new variable matches](https://github.com/serverless/serverless/blob/master/lib/classes/Variables.js#L317).
\</aside>

Discussion [here](https://github.com/serverless/serverless/issues/4984#issuecomment-390692455) and [here](https://github.com/serverless/serverless/issues/4984#issuecomment-400881332) doesn't quite express my concern that users have a knowledge gap if their variable resolution is indefinitely and silently hung.  To help users identify both what is happening and give them the contextual information that should allow them to debug the circumstance, the promise tracker [sets an interval timer to expire every 2.5 seconds](https://github.com/serverless/serverless/blob/master/lib/classes/PromiseTracker.js#L16) (which is usually far beyond the time required to resolve variables, though is infrequently-normal over latent connections [e.g. trans-Atlantic deployments]) and keeps track of the information that is being waited for.  The hope is that reporting that information will support users in taking no action or action that will fix a blocking problem (e.g. missing or the wrong credentials to resolve an SSM variable).

That written, the direct answer to your question about the use case is informing users under a hard to understand class of variable resolution failures.  I have not experienced the latency driven use case, that came from the community.  I have experienced an improper cyclic reference and was helped to resolve it by this messaging.  These failures haven't disappeared and as such I believe support for knowing about and addressing them is important regardless of whether the promise tracker is used.  As I started by admitting the completion message being written during any variable resolution seems uselessly noisy.  The introduction of the completion message was to address [concerns of users about the progress reports](https://github.com/serverless/serverless/issues/4984#issuecomment-545388840).  To avoid the noise while retaining the value I can simply only print the final report if a progress report was previously printed.  This is a simple change that would eliminate the noise while providing the reassurance of successful completion that might otherwise be in question if a progress report is printed.

Two questions:
1. Does this seem like a use case to be concerned with?
2. Is this a solution you can agree with?

I will open the PR with that fix but am happy to be open to alternatives.

**Comment by medikoo** [comment]
Great thanks @erikerikson for clarification.

So aside of serving as memoizer and status indicator (helpful in prolonged resolution issues), promise tracker also allows to handle deep self references in reliable way (which indeed cannot be solved with plain promise constructs).

That indeed justifies this sophisticated resolver.

> Is this a solution you can agree with?

Absolutely, I think we should take this in, just with additional patch you've proposed

**Comment by erikerikson** [comment]
Almost got it.  I got into the weeds a bit hoping the context would help with instancing.  The deep references are not supported by the PromiseTracker.  It does memoize and adds status indication to the promises, a timer for minding their status, as well as reports to print potentially hung promises when that timer triggers.  Now it contingently reports final status.

Thanks for thinking about this deeply.
